designcompose/src/main/java/com/android/designcompose/squoosh/DESIGN.md

```diff
@@ -0,0 +1,92 @@
+3. Submit the request to JNI layout using `SquooshLayout.doLayout`
+4. Put the results of the layout into the `SquooshResolvedNode` tree so that rendering is easier
+   using `populateComputedLayout`
+5. If there are transitions, then generate a new `InteractionState` (and soon `CustomizationContext`)
+   with the transition values applied (these state objects start out without the action of the
+   transition applied, then we ask them to make a copy with all active transitions applied -- so it
+   is the state of the world once all transitions are complete). We then run through steps 1-4 again
+   with the updated context.
+    * This gives us a `SquooshResolvedNode` tree of the world after all transitions have ended.
+    * We use `newSquooshAnimate` to create `SquooshAnimationControl` objects that control progress
+      for each transition. `newSquooshAnimate` identifies similar nodes and different nodes and
+      creates objects that will animate the appropriate property for each. Some nodes from the source
+      tree get moved to the destination tree so that they can be faded out.
+6. If there are transitions, we launch an effect to run an animation clock and update the state of
+   each transition's animation. This is how we can seemingly run many different animations at once
+   -- they all just show different positions between the two trees, and can't be nested.
+7. We emit a `Layout` Composable that occupies the appropriate space and has a `squooshRender`
+   Modifier attached.
+
+## Nested Composables and input events
+Squoosh still uses Compose extensively! Child Composables (including `AndroidView`) can be added,
+and this is supported by having `resolveVariantsRecursively` generate a list of child Composables
+that should be emitted (including externally supplied child Composables, and anything in the
+DesignCompose tree that needs input).
+
+Inside of the `Layout` Composable in `SquooshRoot`, we nest any Child Composables from our list.
+
+Painting just one child Composable is tricky. Compose only lets us paint *all* child Composables in
+one go. So we have a horrible hack where a boxed object is shared between the child Composables and
+the `squooshRender` Modifier. When `squooshRender` wants to paint a particular child Composable it
+updates the boxed object (a `SquooshRenderSelector`) to have the id of the node that should be
+rendered. All of the child Composables have a Modifier that customizes drawing to completely skip
+drawing unless the id in the `SquooshRenderSelector` matches. Yuck! But good enough...
+
+## Animated transitions and enum-driven variant changes
+Squoosh implements animated transitions, and can also animate variant changes caused by enum values
+changing.
+
+This is described above in the "typical flow". When there are animations to be run, they are always
+between states of some kind. We generate two trees, merge parts of them that need to be animated
+between, and then do 
```

Comment:
🤔

Reply:
Missing a work at the end of the last sentence?

Reply:
Cleaned this up.